docs/is_thread_safe/README.md

```diff
@@ -0,0 +1,132 @@
+# Is Stretch Body thread safe?
+
+The short answer: no. This tutorial builds up to a few kinds of threading experiments that prove this, and hopefully will serve as a launching point for making this library thread-safe.
+
+First, let's define what it means to be thread-safe. If we create an instance of the `Robot` class, we'd like to know if we can safely use it in Python threads. Using the `Robot`, there's two ways we interact with the underlying hardware: reading sensor data and writing motion commands. Our experiments will cover:
+
+ 1. Single threaded reading and writing
+ 2. Single writer, but multithreaded readers
+ 3. Multithreading for both reading and writing
+
+Having multiple readers, even if only one thread can send commands, would be useful for visualization purposes.
+
+## Writing
+
+Let's create an instance of the `Robot`:
+
+```python
+import stretch_body.robot
+
+r = stretch_body.robot.Robot()
+assert r.startup()
+assert r.is_homed()
+```
+
+Then create a "writer", that will send motion commands:
+
+```python
+import random
+
+def write():
+    r.lift.move_to(random.uniform(0.3, 1.1))
+    r.arm.move_to(random.uniform(0.0, 0.4))
+    r.end_of_arm.move_to('wrist_roll', random.uniform(-1.0, 1.0))
+    r.head.move_to('head_pan', random.uniform(-1.0, 1.0))
+    r.push_command()
+```
+
+Then let's define a writing-only runner, that sends motion commands at 10hz:
+
+```python
+def wo_runner():
+    """write-only runner"""
+    while True:
+        write()
+        time.sleep(0.1)
+```
+
+### Results
+
+We can run `wo_runner()` in the main thread without any issues. But if we launch `wo_runner()` in a thread:
+
+```python
+import threading
+threading.Thread(target=wo_runner).start()
+```
+
+The program exits with no error. Likely an exception is being thrown in the thread, but isn't being surfaced to the user. Surfacing this exception is the first step towards making this library thread-safe.
```

Comment:
Nice Catch of this in this tutorial, would like investigate as it looks like the custom Exception hooks are not working the way it should in the main thread.

Reply:
I wonder if a Runtime error is being raised, but intentionally I and @aedsinger  decided to mask this out in the exception hook because run-time errors felt common in async event loops that time. Will have to revisit these decisions.
https://github.com/hello-robot/stretch_body/blob/master/body/stretch_body/robot.py#L267

Reply:
Definitely worth trying. I'm a little wary about the custom exception hook and exception filtering we have in Stretch Body right now. In my experience, messing with the default exception behavior yields unexpected results.

However, I'm not familiar with the context that led to this, so I'm sure you have a good reason. It would be great if you could add docs that explain the design decision behind why Stretch Body has custom hooks / filtering, and how it's intended to work.

Lastly, I've seen a lot of `ThreadServiceExits` in Stretch's ROS2 driver recently, and I can't recall why we added it to Stretch Body, and what it's supposed to be doing.

Reply:
Here is the notes I had shared during the time of implementing the Exception hooks during S3 launch. It looks like I was addressing this issue https://github.com/hello-robot/stretch_body/issues/210 where Asyncio was causing trouble to exit cleanly because the even after main thread exit the event loop threads used by asyncio was running (from Transport layer) and causing the stall.

> - Using the [threading.exceptionhook](https://docs.python.org/3/library/threading.html#threading.excepthook) feature to cleanly handle uncaught exceptions on all running threads and exit out of multithreaded applications. I have used this methodology to resolve the [Robot Stalls on exit issue](https://github.com/hello-robot/stretch_body/issues/210) and the reason was uncaught exceptions from running threads that were holding to execute aioserial’s async functions in the transport layer. (still doing tests before merging to master)
> - The threading library's Exception hook allows you to set a custom callback method to handle the exceptions that are usually uncaught in the running threads. I created a custom exception hook to catch these exceptions in the caller’s main thread.